File: board/ownermap.lua
--
-- Copyright (c) 2016-present, Facebook, Inc.
-- All rights reserved.
--
-- This source code is licensed under the BSD-style license found in the
-- LICENSE file in the root directory of this source tree. An additional grant 
-- of patent rights can be found in the PATENTS file in the same directory.
-- 

local ffi = require 'ffi'
local pl = require 'pl.import_into'()
local utils = require('utils.utils')
local common = require("common.common")
local goutils = require('utils.goutils')
local board = require('board.board')

local script_path = common.script_path()
local symbols, s = utils.ffi_include(paths.concat(script_path, "ownermap.h"))
-- local symbols, s = utils.ffi_include(paths.concat(common.lib_path, "board/ownermap.h"))
local C = ffi.load(paths.concat(script_path, "../libs/libownermap.so"))
-- local C = ffi.load("libexperimental_deeplearning_yuandong_go_board_default_policy_c.so")
local om = {}

-- S_DEAD = 8
-- om.dead_white = 8 + common.white
-- om.dead_black = 8 + common.black
-- om.dame_empty = 3
om.dead_white = tonumber(symbols.S_DEAD) + common.white
om.dead_black = tonumber(symbols.S_DEAD) + common.black
om.dame_empty = tonumber(symbols.S_DAME)

-- print("Dead white = " .. om.dead_white)
-- print("Dead white = " .. om.dead_black)
-- print("Dame empty = " .. om.dame_empty)

-- Utilities for Ownermap
function om.new()
    return C.InitOwnermap()
end

function om.free(ownermap)
    C.FreeOwnermap(ownermap)
end

function om.clear_ownermap(ownermap)
    C.ClearOwnermap(ownermap)
end

function om.accu_ownermap(ownermap, board_after_dp)
    C.AccuOwnermap(ownermap, board_after_dp);
end

function om.get_ownermap(ownermap, ratio)
    local ownermap = torch.CharTensor(common.board_size, common.board_size)
    C.GetOwnermap(ownermap, ratio, ownermap:data())
    return ownermap
end

function om.get_deadstones(ownermap, b, ratio)
    local livedead = torch.CharTensor(common.board_size, common.board_size)
    C.GetDeadStones(ownermap, b, ratio, livedead:data(), nil)
    return livedead
end

function om.get_deadlist(livedead)
    -- From livedead we can get back the location of deadstones
    local livedead2 = livedead:view(-1)
    local dead_whites = { }
    local dead_whites_str = { }
    local dead_blacks = { }
    local dead_blacks_str = { }
    for i = 1, common.board_size * common.board_size do
        local x, y = goutils.moveIdx2xy(i)
        local s = goutils.compose_move_gtp(x, y)

        if livedead2[i] == om.dead_white then
            table.insert(dead_whites, {x, y})
            table.insert(dead_whites_str, s)
        elseif livedead2[i] == om.dead_black then
            table.insert(dead_blacks, {x, y})
            table.insert(dead_blacks_str, s)
        end
    end
    return {
        b = dead_blacks,
        w = dead_whites,
        b_str = dead_blacks_str,
        w_str = dead_whites_str,
        dames = dames,
        dames_str = dames_str
    }
end

function om.get_territorylist(territory) 
    local territory2 = territory:view(-1)
    local whites = { }
    local whites_str = { } 
    local blacks = { }
    local blacks_str = { } 
    local dames = { }
    local dames_str = { }
    for i = 1, common.board_size * common.board_size do
        local x, y = goutils.moveIdx2xy(i)
        local s = goutils.compose_move_gtp(x, y)

        if territory2[i] == common.white then
            table.insert(whites, {x, y})
            table.insert(whites_str, s)
        elseif territory2[i] == common.black then
            table.insert(blacks, {x, y})
            table.insert(blacks_str, s)
        elseif territory2[i] == om.dame_empty then
            -- Dame
            table.insert(dames, {x, y})
            table.insert(dames_str, s)
        end
    end
    return {
        b = blacks,
        w = whites,
        b_str = blacks_str,
        w_str = whites_str,
        dames = dames,
        dames_str = dames_str
    }
end

function om.get_ownermap_float(ownermap_ptr, player)
    local ownermap = torch.FloatTensor(common.board_size, common.board_size)
    C.GetOwnermapFloat(ownermap_ptr, player, ownermap:data())
    return ownermap
end

function om.show_deadstones(b, stones)
    C.ShowDeadStones(b, stones:data())
end

function om.show_stones_prob(ownermap, player)
    C.ShowStonesProb(ownermap, player)
end

function om.get_ttscore_ownermap(ownermap, b)
    local livedead = torch.CharTensor(common.board_size, common.board_size)
    local territory = torch.CharTensor(common.board_size, common.board_size)
    local score = C.GetTTScoreOwnermap(ownermap, b, livedead:data(), territory:data())
    return score, livedead, territory
end

function om.print_list(t)
    local s = "" 
    for _, c in ipairs(t) do
        s = s .. c .. " "
    end
    print(s)
end

function om.util_compute_final_score(ownermap, b, komi, trial, def_policy_func)
     local new_ownermap
     if not ownermap then
         ownermap = om.new()
         new_ownermap = true
     end
     assert(ownermap)
     assert(def_policy_func)

     trial = trial or 1000
     komi = komi or 6.5

     om.clear_ownermap(ownermap)
     local scores = torch.Tensor(trial)
     for i = 1, trial do
         local b2 = board.copyfrom(b)
         scores[i] = def_policy_func(b2, -1) - komi
         om.accu_ownermap(ownermap, b2)
     end
     local score, livedead, territory = om.get_ttscore_ownermap(ownermap, b)
     
     --[[
     local territorylist =  om.get_territorylist(territory) 
     print(string.format("#W = %d, #B = %d, #dame = %d", 
                #territorylist.w, #territorylist.b, #territorylist.dames))
     
     print("White = ")
     print_list(territorylist.w_str)
     print("Black = ")
     print_list(territorylist.b_str)
     print("Dame = ")
     print_list(territorylist.dames_str)
 
     print("Score before komi = " .. score)
     ]]

     if new_ownermap then
         om.free(ownermap)
     end
     return score - komi, livedead, territory, scores
end

-- Some utilities for dp
function om.compute_stats(scores)
    if scores == nil then
        return { 
            std = 0,
            max = 0,
            min = 0,
            mean = 0,
            advantage = '-'
        }
    else
        if type(scores) == 'table' then
            scores = torch.FloatTensor(scores)
        end
        return {
            mean = scores:mean(),
            std = scores:std(),
            max = scores:max(),
            min = scores:min(),
            advantage = scores:mean() > 0 and 'B' or 'W'
        }
    end
end

return om

File: board/ownermap.c
//
// Copyright (c) 2016-present, Facebook, Inc.
// All rights reserved.
//
// This source code is licensed under the BSD-style license found in the
// LICENSE file in the root directory of this source tree. An additional grant 
// of patent rights can be found in the PATENTS file in the same directory.
// 

#include "ownermap.h"
// Ownermap
typedef struct {
  // Ownermap
  int total_ownermap_count;
  // Histogram. S_EMPTY (S_UNKNOWN), S_BLACK, S_WHITE, S_OFF_BOARD (S_DAME)
  int ownermap[MACRO_BOARD_SIZE][MACRO_BOARD_SIZE][4];
} Handle;

void *InitOwnermap() {
  Handle *hh = (Handle *)malloc(sizeof(Handle));
  if (hh == NULL) error("Ownermap handle cannot be initialized.");
  return hh;
}

void FreeOwnermap(void *hh) {
  free(hh);
}

void ClearOwnermap(void *hh) {
  Handle *h = (Handle *)hh;
  // Ownermap
  h->total_ownermap_count = 0;
  memset(h->ownermap, 0, sizeof(h->ownermap));
}

void AccuOwnermap(void *hh, const Board *board) {
  // Accumulate the ownermap with the board situation.
  // Usually the current board situation is after the default policy is applied.
  Handle *h = (Handle *)hh;
  // Accumulate the ownermap
  for (int i = 0; i < BOARD_SIZE; ++i) {
    for (int j = 0; j < BOARD_SIZE; ++j) {
      Coord c = OFFSETXY(i, j);
      Stone s = board->_infos[c].color;
      if (s == S_EMPTY) {
        s = GetEyeColor(board, c);
      }
      h->ownermap[i][j][s] ++;
    }
  }
  h->total_ownermap_count ++;
}

float OwnermapFloatOne(Handle *h, int i, int j, Stone player) {
  return ((float) h->ownermap[i][j][player]) / h->total_ownermap_count;
}

Stone OwnermapJudgeOne(Handle *h, int i, int j, float ratio) {
  int empty = h->ownermap[i][j][S_EMPTY];
  int black = h->ownermap[i][j][S_BLACK];
  int white = h->ownermap[i][j][S_WHITE];
  int n = h->total_ownermap_count;

  int thres = (int)(n * ratio);

  if (empty >= thres) return S_DAME;
  if (empty + black >= thres) return S_BLACK;
  if (empty + white >= thres) return S_WHITE;
  return S_UNKNOWN;
}

void GetDeadStones(void *hh, const Board *board, float ratio, Stone *livedead, Stone *group_stats) {
  // Threshold the ownermap and determine.
  Handle *h = (Handle *)hh;

  Stone *internal_group_stats = NULL;

  if (group_stats == NULL) {
    internal_group_stats = (Stone *)malloc(board->_num_groups * sizeof(Stone));
    group_stats = internal_group_stats;
  }

  memset(group_stats, S_EMPTY, board->_num_groups * sizeof(Stone));

  for (int i = 0; i < BOARD_SIZE; ++i) {
    for (int j = 0; j < BOARD_SIZE; ++j) {
      Coord c = OFFSETXY(i, j);
      Stone s = board->_infos[c].color;
      Stone owner = OwnermapJudgeOne(h, i, j, ratio);

      // printf("owner at (%d, %d) = %d\n", i, j, owner);
      short id = board->_infos[c].id;
      if (owner == S_UNKNOWN) {
        group_stats[id] = s | S_UNKNOWN;
      } else if (! (group_stats[id] & S_UNKNOWN)) {
        // The group has deterministic state or empty.
        Stone stat = s;
        if (owner == s) stat |= S_ALIVE;
        else if (owner == OPPONENT(s)) stat |= S_DEAD;
        else stat |= S_UNKNOWN;

        if (group_stats[id] == S_EMPTY) group_stats[id] = stat;
        else if (group_stats[id] != stat) group_stats[id] = s | S_UNKNOWN;
      }
    }
  }
  // Once we get the group stats, we thus can fill the ownermap.
  if (livedead != NULL) {
    // Zero out everything else.
    memset(livedead, S_EMPTY, BOARD_SIZE * BOARD_SIZE * sizeof(Stone));
    for (int i = 1; i < board->_num_groups; ++i) {
      TRAVERSE(board, i, c) {
        livedead[EXPORT_OFFSET(c)] = group_stats[i];
      } ENDTRAVERSE
    }
  }

  if (internal_group_stats != NULL) free(internal_group_stats);
}

void GetOwnermap(void *hh, float ratio, Stone *ownermap) {
  // Threshold the ownermap and determine.
  Handle *h = (Handle *)hh;

  for (int i = 0; i < BOARD_SIZE; ++i) {
    for (int j = 0; j < BOARD_SIZE; ++j) {
      Coord c = OFFSETXY(i, j);
      ownermap[EXPORT_OFFSET(c)] = OwnermapJudgeOne(h, i, j, ratio);
    }
  }
}

void GetOwnermapFloat(void *hh, Stone player, float *ownermap) {
  Handle *h = (Handle *)hh;
  for (int i = 0; i < BOARD_SIZE; ++i) {
    for (int j = 0; j < BOARD_SIZE; ++j) {
      Coord c = OFFSETXY(i, j);
      ownermap[EXPORT_OFFSET(c)] = OwnermapFloatOne(h, i, j, player);
    }
  }
}

float GetTTScoreOwnermap(void *hh, const Board *board, Stone *livedead, Stone *territory) {
  Stone group_stats[MAX_GROUP];
  GetDeadStones(hh, board, 0.5, livedead, group_stats);
  return GetTrompTaylorScore(board, group_stats, territory);
}

void ShowDeadStones(const Board *board, const Stone *stones) {
  // Show the board with ownership
  char buf[2000];
  int len = 0;
  len += sprintf(buf + len, "   A B C D E F G H J K L M N O P Q R S T\n");
  char stone[3];
  stone[2] = 0;
  for (int j = BOARD_SIZE - 1; j >= 0; --j) {
    len += sprintf(buf + len, "%2d ", j + 1);
    for (int i = 0; i < BOARD_SIZE; ++i) {
      Coord c = OFFSETXY(i, j);
      Stone s = board->_infos[c].color;
      if (HAS_STONE(s)) {
        char ss = (s == S_BLACK ? 'X' : 'O');
        // Make it lower case if the stone are dead.
        Stone stat = stones[EXPORT_OFFSET(c)];
        if (stat & S_DEAD) ss |= 0x20;
        stone[0] = ss;
        stone[1] = ( (stat & S_UNKNOWN) ? '?' : (c == board->_last_move ? ')' : ' '));

      } else if (s == S_EMPTY) {
        if (STAR_ON19(i, j))
          strcpy(stone, "+ ");
        else
          strcpy(stone, ". ");
      } else strcpy(stone, "# ");
      len += sprintf(buf + len, stone);
    }
    len += sprintf(buf + len, "%d\n", j + 1);
  }
  len += sprintf(buf + len, "   A B C D E F G H J K L M N O P Q R S T");
  // Finally print
  printf(buf);
}

void ShowStonesProb(void *hh, Stone player) {
  // Show the board with ownership
  char buf[20000];
  int len = 0;
  Handle *h = (Handle *)hh;

  const char *prompt = "   A      B      C      D      E      F      G      H      J      K      L      M      N      O      P      Q      R      S      T\n";
  len += sprintf(buf + len, prompt);
  for (int j = BOARD_SIZE - 1; j >= 0; --j) {
    len += sprintf(buf + len, "%2d ", j + 1);
    for (int i = 0; i < BOARD_SIZE; ++i) {
      Coord c = OFFSETXY(i, j);
      float val = OwnermapFloatOne(h, i, j, player);
      len += sprintf(buf + len, "%.3f  ", val);
    }
    len += sprintf(buf + len, "%d\n", j + 1);
  }
  len += sprintf(buf + len, prompt);
  // Finally print
  printf(buf);
}


